### PR TITLE
feat: foxy withdraw

### LINKED ISSUE #1328: Instant Unstake Should Unstake 100%
When Instant unstake is selected, select 100% of the FOXy tokens in the amount and disable both the input and percentage buttons

### PR DESCRIPTION
**Pull request by DaoDev44**
## Description
NOTE: DO NOT MERGE until [this lib](https://github.com/shapeshift/lib/pull/498) PR is merged.

Add withdrawal functionality to `FOXy`

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1328 

## Risk

The risk associated with this PR is minimal, given this work is behind the Foxy Investor feature flag. Moving`WithdrawalType`  out of web and into the `types` package in `lib` can be associated with some risk if the package isn't bumped, but that is minimal as well.

## Testing

NOTE: In order to go through the Withdrawal flow of FOXy, you must have a foxy position.

- Turn on `FoxyInvestor` feature flag
- Navigate to the `Defi/Earn` page
- Click on the `type` row of the earn screen to sort earn opportunities by type.
- Click on `FOX` earning type
- Click on `Withdrawal` tab

## Screenshots (if applicable)

Screen Shot 2022-03-26 at 8 33 44 PM
Screen Shot 2022-03-26 at 8 33 58 PM
Screen Shot 2022-03-26 at 8 35 27 PM

### CODE REVIEW

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+    return hasValidBalance || 'common.insufficientFunds'
+  }
+
+  const cryptoAmountAvailable = bnOrZero(balance).div(`1e+${asset?.precision}`)
+  const fiatAmountAvailable = bnOrZero(cryptoAmountAvailable).times(marketData.price)
+
+  const renderRoute = (route: { step?: number; path: string; label: string }) => {
+    let statusIcon: React.ReactElement = <ArrowForwardIcon />
+    let statusText = StatusTextEnum.pending
+    if (state.withdraw.txStatus === 'success') {
```
Comment (reallybeard) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:340]: Check out the changes I made to foxy deposit status stuff. We pass in a BG color now -- https://github.com/shapeshift/web/blob/380c1696dfc77132cdbc68ace55249542217881c/src/features/defi/providers/foxy/components/FoxyManager/Deposit/FoxyDeposit.tsx#L356

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+      case WithdrawPath.Status:
+        return (
+          <TxStatus
+            onClose={handleCancel}
+            onContinue={handleViewPosition}
+            loading={state.loading}
+            continueText='modals.status.position'
+            closeText='modals.status.close'
+            statusText={statusText}
+            statusIcon={statusIcon}
```
Comment (reallybeard) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:473]: Here is where you'll put that BG prop -- https://github.com/shapeshift/web/blob/380c1696dfc77132cdbc68ace55249542217881c/src/features/defi/providers/foxy/components/FoxyManager/Deposit/FoxyDeposit.tsx#L524

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+      const _allowance = await api.allowance({
+        tokenContractAddress: rewardId,
+        contractAddress,
+        userAddress: state.userAddress
+      })
+
+      const allowance = bnOrZero(_allowance).div(`1e+${asset.precision}`)
+
+      // Skip approval step if user allowance is greater than requested deposit amount
+      if (allowance.gt(formValues.cryptoAmount)) {
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:197]:
```suggestion
      if (allowance.gte(formValues.cryptoAmount)) {
```

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+          tokenContractAddress: rewardId,
+          contractAddress,
+          userAddress: state.userAddress
+        }),
+        api.getGasPrice()
+      ])
+      return bnOrZero(gasPrice).times(gasLimit).toFixed(0)
+    } catch (error) {
+      console.error('FoxyWithdraw:getApproveEstimate error:', error)
+      toast({
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:239]: Possibly set the estimated gas value in state if the call fails

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+        status: 'error'
+      })
+    } finally {
+      dispatch({ type: FoxyWithdrawActionType.SET_LOADING, payload: false })
+    }
+  }
+
+  const handleConfirm = async () => {
+    try {
+      if (!state.userAddress || !rewardId || !walletState.wallet) return
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:296]: add `loading`

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+  const handleViewPosition = () => {
+    browserHistory.push('/defi')
+  }
+
+  const handleCancel = () => {
+    browserHistory.goBack()
+  }
+
+  const validateCryptoAmount = (value: string) => {
+    const crypto = bnOrZero(balance).div(`1e+${asset.precision}`)
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:342]:
```suggestion
    const crypto = bnOrZero(bn(balance).div(`1e+${asset.precision}`))
```

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+    const _value = bnOrZero(value)
+    const hasValidBalance = crypto.gt(0) && _value.gt(0) && crypto.gte(value)
+    if (_value.isEqualTo(0)) return ''
+    return hasValidBalance || 'common.insufficientFunds'
+  }
+
+  const validateFiatAmount = (value: string) => {
+    const crypto = bnOrZero(balance).div(`1e+${asset.precision}`)
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:350]:
```suggestion
    const crypto = bnOrZero(bn(balance).div(`1e+${asset.precision}`))
```

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+    const fiat = crypto.times(marketData.price)
+    const _value = bnOrZero(value)
+    const hasValidBalance = fiat.gt(0) && _value.gt(0) && fiat.gte(value)
+    if (_value.isEqualTo(0)) return ''
+    return hasValidBalance || 'common.insufficientFunds'
+  }
+
+  const cryptoAmountAvailable = bnOrZero(balance).div(`1e+${asset?.precision}`)
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:358]:
```suggestion
  const cryptoAmountAvailable = bnOrZero(bn(balance).div(`1e+${asset?.precision}`))
```

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+  const fiatAmountAvailable = bnOrZero(cryptoAmountAvailable).times(marketData.price)
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:359]:
```suggestion
  const fiatAmountAvailable = bnOrZero(bn(cryptoAmountAvailable).times(marketData.price))
```

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 10]
```diff
@@ -0,0 +1,651 @@
...
+            asset={asset}
+            feeAsset={feeAsset}
+            cryptoEstimatedGasFee={bnOrZero(state.approve.estimatedGasCrypto)
+              .div(`1e+${feeAsset.precision}`)
+              .toFixed(5)}
+            disableAction
+            fiatEstimatedGasFee={bnOrZero(state.approve.estimatedGasCrypto)
+              .div(`1e+${feeAsset.precision}`)
+              .times(feeMarketData.price)
+              .toFixed(2)}
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:435]:
```suggestion
            fiatEstimatedGasFee={bnOrZero(bn(state.approve.estimatedGasCrypto)
              .div(`1e+${feeAsset.precision}`)
              .times(feeMarketData.price))
              .toFixed(2)}
```

Comment (DaoDev44) [reply src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:435]: same for above

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 11]
```diff
@@ -0,0 +1,651 @@
...
+                  <Text translation='modals.confirm.estimatedGas' />
+                </Row.Label>
+                <Row.Value>
+                  <Box textAlign='right'>
+                    <Amount.Fiat
+                      fontWeight='bold'
+                      value={bnOrZero(state.withdraw.estimatedGasCrypto)
+                        .div(`1e+${feeAsset.precision}`)
+                        .times(feeMarketData.price)
+                        .toFixed(2)}
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:514]:
```suggestion
                      value={bnOrZero(bn(state.withdraw.estimatedGasCrypto)
                        .div(`1e+${feeAsset.precision}`)
                        .times(feeMarketData.price))
                        .toFixed(2)}
```

File: src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx [hunk 12]
```diff
@@ -0,0 +1,651 @@
...
+                    />
+                    <Amount.Crypto
+                      color='gray.500'
+                      value={bnOrZero(state.withdraw.estimatedGasCrypto)
```
Comment (DaoDev44) [inline src/features/defi/providers/foxy/components/FoxyManager/Withdraw/FoxyWithdraw.tsx:518]: same as above

File: src/features/defi/components/Deposit/Deposit.tsx [hunk 13]
```diff
@@ -261,6 +261,29 @@ export const Deposit = ({
...
+                <ButtonGroup width='full' justifyContent='space-between' size='sm' px={4} py={2}>
+                  {percentOptions.map(option => (
+                    <Button
+                      isActive={option === percent}
+                      key={option}
+                      variant='ghost'
+                      colorScheme='blue'
+                      onClick={() => handlePercentClick(option)}
+                    >
+                      {option === 1 ? (
```
Comment (DaoDev44) [inline src/features/defi/components/Deposit/Deposit.tsx:273]: comment // 100%

**Review by cjthompson — APPROVED** [review]

**Review by toshiSat — APPROVED** [review]

**Pull request merged**